src/v/rpc/connection_cache.cc

```diff
@@ -11,62 +11,173 @@
+
+    auto& n2s = _node_to_shards[node];
+
+    for (unsigned i = 0; i < connections_per_node; i++) {
+        auto& cps = _connections_per_shard[i];
+        ret.add_connections.emplace_back(node, std::get<0>(cps));
+        std::get<1>(cps)++;
+        n2s.emplace(std::get<0>(cps));
+    }
+
+    get_shard_to_connection_assignments(ret, node);
+
+    return ret;
+}
+
+connection_allocation_strategy::changes
+connection_allocation_strategy::remove_connection_assignments_for(
+  model::node_id node) {
+    changes ret;
+
+    if (!has_connection_assignments_for(node)) {
+        return ret;
+    }
+
+    auto& conn_shards = _node_to_shards[node];
+
+    for (auto shard : conn_shards) {
+        auto s_it = std::find_if(
+          _connections_per_shard.begin(),
+          _connections_per_shard.end(),
+          [shard](const auto& t) { return std::get<0>(t) == shard; });
+
+        if (s_it != _connections_per_shard.cend()) {
+            std::get<1>(*s_it)--;
+            ret.remove_connections.emplace_back(node, std::get<0>(*s_it));
+        }
+    }
+
+    _node_to_shards.erase(node);
+
+    for (auto shard : std::views::iota((unsigned)0, _total_shards)) {
+        ret.remove_mapping.emplace_back(node, shard);
+    }
+
+    return ret;
+}
+
+void connection_allocation_strategy::get_shard_to_connection_assignments(
+  changes& c, model::node_id node) {
+    auto& conn_shards = _node_to_shards.at(node);
+
+    // Assign shards that own a connection to use their own connection
+    for (auto shard : conn_shards) {
+        c.add_mapping.emplace_back(node, shard, shard);
+    }
+
+    // Randomly assign the remaining shards to connections.
+    std::vector<ss::shard_id> conn_shards_r{
+      conn_shards.begin(), conn_shards.end()};
+    std::shuffle(conn_shards_r.begin(), conn_shards_r.end(), _g);
+
+    auto j = 0;
+    for (auto shard : std::views::iota((unsigned)0, _total_shards)) {
+        if (conn_shards.contains(shard)) {
+            continue;
+        }
+
+        auto connection_idx = j++ % conn_shards_r.size();
+        c.add_mapping.emplace_back(node, shard, conn_shards_r[connection_idx]);
+    }
+}
+
 connection_cache::connection_cache(
   ss::sharded<ss::abort_source>& as,
-  std::optional<connection_cache_label> label)
+  std::optional<connection_cache_label> label,
+  unsigned connections_per_node)
   : _label(std::move(label)) {
     _as_subscription = as.local().subscribe(
       [this]() mutable noexcept { shutdown(); });
+    if (ss::this_shard_id() == _coordinator_shard) {
+        _coordinator_state = std::make_unique<coordinator_state>(
+          mutex(),
+          connection_allocation_strategy(connections_per_node, ss::smp::count));
+    }
+}
+
+std::optional<ss::shard_id> connection_cache::shard_for(
+  model::node_id, ss::shard_id, model::node_id n, ss::shard_id) const {
+    if (!_connection_map.contains(n)) {
+        return {};
+    }
+    return {_connection_map.at(n)};
```

Comment:
Use `find` for this pattern? Avoids an extra hash table look up.

Reply:
Right you are, switching over to `find`